PerfDataExtension/PerfDataEvent.cs

```diff
@@ -0,0 +1,793 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Performance.Toolkit.Plugins.PerfDataExtension
+{
+    using Microsoft.LinuxTracepoints;
+    using Microsoft.LinuxTracepoints.Decode;
+    using Microsoft.Performance.SDK.Extensibility;
+    using System;
+    using System.Diagnostics.Tracing;
+    using Debug = System.Diagnostics.Debug;
+    using Timestamp = Microsoft.Performance.SDK.Timestamp;
+
+    /// <summary>
+    /// Stores and provides access to an event from a perf.data file.
+    /// Stores the event's header, raw data bytes, and core offsets/lengths.
+    /// Other information is extracted on-demand by properties.
+    /// </summary>
+    public sealed class PerfDataEvent : IKeyedDataType<PerfEventHeaderType>
+    {
+        private readonly byte[] contents;
+        private readonly PerfEventDesc eventDesc;
+        private readonly ulong fileRelativeTime;
+        private readonly PerfEventHeader header;
+
+        private readonly ushort topLevelFieldCount;
+        private readonly PerfByteReader byteReader;
+
+        // The following fields are zero for non-Sample events.
+
+        private readonly byte activityIdLength; // EventHeader only. 0, 16, or 32.
+        private readonly ushort activityIdStart; // EventHeader only. Offset into contents for activity ID + related ID.
+        private readonly ushort rawDataLength;
+        private readonly ushort rawDataStart;
+        private readonly ushort eventHeaderNameLength; // EventHeader only. Length of the EventHeader event name.
+        private readonly ushort eventHeaderNameStart; // EventHeader only. Offset into contents for the EventHeader event name.
+
+        /// <summary>
+        /// For raw events, i.e. non-Sample events with no event info.
+        /// </summary>
+        public PerfDataEvent(
+            PerfByteReader byteReader,
+            in PerfEventBytes bytes,
+            ulong fileRelativeTime)
+        {
+            var bytesSpan = bytes.Span;
+            var hasContent = bytesSpan.Length > 8;
+            Debug.Assert(bytesSpan.Length >= 8);
+
+            this.contents = hasContent ? bytesSpan.Slice(8).ToArray() : Array.Empty<byte>();
+            this.eventDesc = PerfEventDesc.Empty;
+            this.fileRelativeTime = fileRelativeTime;
+            this.header = bytes.Header;
+            this.topLevelFieldCount = (ushort)(hasContent ? 1 : 0); // If there is content, we'll make a "raw" field.
+            this.byteReader = byteReader;
+        }
+
+        /// <summary>
+        /// For non-Sample events with event info.
+        /// </summary>
+        public PerfDataEvent(

```

Comment:
FYI - cpu-clock events don't seem to be decoded today, but that is probably expected

Reply:
Not sure which specific events you're talking about, but yeah, the library currently only decodes tracepoints. Non-tracepoint events are just shown as "raw". It's an open question about how non-tracepoint events should be handled -- should they be hidden (only used by cookers/tables, never directly exposed to the user) or should some of them get hard-coded decodings (they don't have any runtime metadata -- they're just "well-known" events).